shared/mputil/scatter.go

```diff
@@ -0,0 +1,77 @@
+package mputil
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+)
+
+// Scatter scatters a computation across multiple goroutines.
+// This breaks the task in to a number of chunks and executes those chunks in parallel with the function provided.
+// Results returned are collected and presented a a set of WorkerResults, which can be reassembled by the calling function.
+// Any error that occurs in the workers will be passed back to the calling function.
+func Scatter(inputLen int, sFunc func(int, int, *sync.Mutex) (interface{}, error)) ([]*WorkerResults, error) {
+	if inputLen <= 0 {
+		return nil, errors.New("input length must be greater than 0")
+	}
+
+	chunkSize := calculateChunkSize(inputLen)
+	workers := inputLen / chunkSize
+	if inputLen%chunkSize != 0 {
+		workers++
+	}
+	resultCh := make(chan *WorkerResults, workers)
+	defer close(resultCh)
+	errorCh := make(chan error, workers)
+	defer close(errorCh)
+	mutex := new(sync.Mutex)
```

Comment:
maybe lets use a RWMutex here instead ? in the future we might want multiple readers for a certain portion of a function

Reply:
Updated.